### PR TITLE
Implement takeUpTo

### PR DESCRIPTION
**Pull request by afsalthaj**
This is the initial commit on the implementation of takeUpTo. Please review.

### CODE REVIEW

**Comment by afsalthaj** [comment]
@wi101  I just raised an initial commit on takeUpTo.  Could you please review and see if it makes sense ?

**Comment by wi101** [comment]
PR for #7 (@afsalthaj I wrote this comment to link this PR with the issue)

**Comment by afsalthaj** [comment]
@wi101 Thanks for tagging with the issue.
 Looking forward to your comments.

**Comment by wi101** [comment]
@afsalthaj I added some review comments

**Comment by afsalthaj** [comment]
@wi101 It seems I am not able to view the comments.

**Comment by wi101** [comment]
:(

**Comment by wi101** [comment]
I am sorry about that..

File: src/main/scala/scalaz/ioqueue/Queue.scala [hunk 1]
```diff
@@ -84,6 +84,29 @@ class Queue[A] private (capacity: Int, ref: Ref[State[A]]) {
...
+    ref.modify[List[A]] {
+      case Surplus(values, putters) if max > values.size =>
+        (values.toList, Surplus(IQueue.empty[A], putters))
+      case Surplus(values, putters) =>
+        val (listA, queue) =
+          (0 until max)
+            .foldLeft((Nil: List[A], values)) {
+              case ((l, q), _) =>
+                q.dequeueOption match {
+                  case Some((a, b)) => (a :: l, b)
```
[the remark this answers is not in the document]
Comment (afsalthaj) [reply src/main/scala/scalaz/ioqueue/Queue.scala]: @wi101  Doesn't it have performance impact? The list append is linear asymptotically? This is why I chose to have append to head and reverse once and for all.

Ex: `(0 to 50000).foldLeft(List[Int]())((a, b) => b :: a).reverse` should be faster than
`(0 to 50000).foldLeft(List[Int]())((a, b) => a :+ b)`?

[the remark this answers is not in the document]
Comment (jdegoes) [reply src/main/scala/scalaz/ioqueue/Queue.scala]: How about just calling `queue.take(max)`. 😄

[the remark this answers is not in the document]
Comment (afsalthaj) [reply src/main/scala/scalaz/ioqueue/Queue.scala]: Worked. Completely missed out those functionalities in queue 😄

```
  final def takeUpTo(max: Int): IO[Nothing, List[A]] =
    ref.modify[List[A]] {
      case Surplus(values, putters) =>
        val listA = values.take(max).toList
        val queue = values.drop(max)

        (listA, Surplus(queue, putters))
      case state @ Deficit(_) => (Nil, state)
    }
```

File: src/test/scala/scalaz/ioqueue/QueueSpec.scala [hunk 2]
```diff
@@ -142,6 +164,79 @@ class QueueSpec(implicit ee: ExecutionEnv) extends AbstractRTSSpec with AroundTi
...
+      queue <- Queue.bounded[Int](100)
+      _     <- queue.offer(10)
+      _     <- queue.offer(20)
+      _     <- queue.offer(30)
+      _     <- queue.offer(40)
+      list1 <- queue.takeUpTo(2)
+      list2 <- queue.takeUpTo(2)
+    } yield (list1, list2) must_=== ((List(10, 20), List(30, 40)))
+  )
+
```
[the remark this answers is not in the document]
Comment (afsalthaj) [reply src/test/scala/scalaz/ioqueue/QueueSpec.scala]: Added these tests 👍

**Comment by afsalthaj** [comment]
@wi101  Addressed your PR comments, and added @jdegoes 's suggestion..

**Comment by wi101** [comment]
@afsalthaj Good job 🎉

**Pull request merged**